lib/agent.js

```diff
@@ -0,0 +1,81 @@
+'use strict'
+/* global WeakRef, FinalizationRegistry */
+
+const { Pool } = require('../')
+
+const kAgentPoolOpts = Symbol('agent pool opts')
+const kAgentCache = Symbol('agent cache')
+const kAgentRegistry = Symbol('agent registry')
+class Agent {
+  constructor (opts) {
+    this[kAgentPoolOpts] = opts
+    this[kAgentCache] = new Map()
+    this[kAgentRegistry] = new FinalizationRegistry(key => {
+      // get the Pool WeakRef from the cache
+      const ref = this[kAgentCache].get(key)
+      // if the WeakRef exists and the object has been reclaimed
+      // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
```

Comment:
I actually think the second call is necessary as there is a possible ref race otherwise. See the Chrome page on WeakRefs.

Reply:
https://github.com/tc39/proposal-weakrefs

Reply:
Ooh I also realized me putting all of this in the class is wrong too. V8 page on WeakRefs is very in depth